internal/framework/fetch/fetch.go

```diff
@@ -0,0 +1,320 @@
+//
+//counterfeiter:generate . Fetcher
+type Fetcher interface {
+	GetRemoteFile(targetURL string) ([]byte, error)
+}
+
+// DefaultFetcher is the default implementation of Fetcher.
+type DefaultFetcher struct {
+	httpClient       *http.Client
+	checksumLocation string
+	retryBackoff     RetryBackoffType
+	timeout          time.Duration
+	retryMaxDelay    time.Duration
+	retryAttempts    int32
+	checksumEnabled  bool
+}
+
+// NewDefaultFetcher creates a new DefaultFetcher.
+func NewDefaultFetcher(opts ...Option) *DefaultFetcher {
+	fetcher := &DefaultFetcher{
+		httpClient: &http.Client{
+			Timeout: defaultTimeout,
+		},
+		timeout:       defaultTimeout,
+		retryAttempts: defaultRetryAttempts,
+		retryMaxDelay: defaultRetryMaxDelay,
+		retryBackoff:  RetryBackoffExponential,
+	}
+
+	for _, opt := range opts {
+		opt(fetcher)
+	}
+
+	return fetcher
+}
+
+// GetRemoteFile fetches a remote file with retry logic and optional validation.
+func (f *DefaultFetcher) GetRemoteFile(targetURL string) ([]byte, error) {
+	ctx := context.Background()
+
+	if !strings.HasPrefix(targetURL, "http://") && !strings.HasPrefix(targetURL, "https://") {
+		return nil, fmt.Errorf("unsupported URL scheme (supported: http://, https://)")
+	}
+
+	backoff := createBackoffConfig(f.retryBackoff, f.retryAttempts, f.retryMaxDelay)
+	var lastErr error
+	var result []byte
+
+	err := wait.ExponentialBackoffWithContext(ctx, backoff, func(ctx context.Context) (bool, error) {
+		data, err := f.getFileContent(ctx, targetURL, f.timeout)
+		if err != nil {
+			lastErr = fmt.Errorf("HTTP error for %s: %w", targetURL, err)
+
+			shouldRetry, retryErr := f.shouldRetryHTTPError(err)
+			if !shouldRetry {
+				return false, retryErr
+			}
+			return false, nil
+		}
+
+		if f.checksumEnabled {
+			if err := f.validateChecksum(ctx, data, targetURL); err != nil {
+				lastErr = fmt.Errorf("checksum validation failed: %w", err)
+
+				var checksumErr *ChecksumMismatchError
+				if errors.As(err, &checksumErr) {
+					return false, lastErr // Stop retrying on checksum mismatches
+				}
+
+				if strings.Contains(err.Error(), "failed to fetch checksum from") {
+					shouldRetry, retryErr := f.shouldRetryHTTPError(err)
+					if !shouldRetry {
+						return false, retryErr
+					}
+					return false, nil
+				}
+
+				return false, nil // Retry on other checksum errors
+			}
+		}
+
+		result = data
+		return true, nil
+	})
+	if err != nil {
+		// If the backoff timed out or was aborted by a non-retryable error,
+		// return the last recorded error for better context.
+		if lastErr != nil {
+			return nil, lastErr
+		}
+		return nil, fmt.Errorf("failed to fetch HTTP file after retries: %w", err)
+	}
+
+	if result != nil {
+		return result, nil
+	}
+
+	// This case should ideally not be reached, but as a fallback, return the last known error.
+	if lastErr != nil {
+		return nil, lastErr
+	}
+
+	return nil, fmt.Errorf("failed to fetch HTTP file %s: unknown error", targetURL)
```

Comment:
I think this all could just be simplified to

```
return result, lastErr
```

Reply:
You're right, this can be simplified, but to `return result, nil` (as `lastErr` could hold erroneous errors from previously failed iterations that eventually succeeded)